# EXERCICI 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [24]:
import pandas as pd
airlines_delay = pd.read_csv('/Users/angel/Desktop/DelayedFlights.csv')
airlines_delay

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

In [25]:
duplicados = airlines_delay[airlines_delay.duplicated()]
print(duplicados)

Empty DataFrame
Columns: [Unnamed: 0, Year, Month, DayofMonth, DayOfWeek, DepTime, CRSDepTime, ArrTime, CRSArrTime, UniqueCarrier, FlightNum, TailNum, ActualElapsedTime, CRSElapsedTime, AirTime, ArrDelay, DepDelay, Origin, Dest, Distance, TaxiIn, TaxiOut, Cancelled, CancellationCode, Diverted, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay]
Index: []

[0 rows x 30 columns]


No hay duplicados

### Valores NaN.

In [26]:
print(airlines_delay.isnull().sum())

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64


Selecciono las filas con valores NaN en AirTime y ArrDelay

In [29]:
filas_nan = airlines_delay[airlines_delay['AirTime'].isnull() & airlines_delay['ArrDelay'].isnull()]
filas_nan

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1280           1763  2008      1           3          4    922.0         915   
1372           1911  2008      1           3          4   2325.0        1900   
1776           2651  2008      1           4          5   1949.0        1905   
1831           2726  2008      1           4          5    737.0         705   
2244           3672  2008      1           4          5   1849.0        1630   
...             ...   ...    ...         ...        ...      ...         ...   
1935651     7006289  2008     12          10          3   1459.0        1447   
1935716     7006401  2008     12          11          4   1355.0        1106   
1935876     7006809  2008     12          11          4   1026.0         955   
1935978     7007034  2008     12          11          4   1527.0        1520   
1936470     7008584  2008     12          12          5    703.0         630   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1280         NaN        1050            WN  ...     NaN    12.0          0   
1372         NaN        2030            WN  ...     NaN    11.0          0   
1776         NaN        1910            WN  ...     NaN    11.0          0   
1831         NaN         825            WN  ...     NaN    13.0          0   
2244         NaN        1755            WN  ...     NaN    12.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1935651      NaN        1650            DL  ...     NaN    37.0          1   
1935716      7.0        1950            DL  ...    13.0    17.0          0   
1935876      NaN        1219            DL  ...     NaN     NaN          1   
1935978   2106.0        1708            DL  ...     9.0    19.0          0   
1936470      NaN         734            DL  ...     NaN    33.0          1   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1280                    N         1           NaN           NaN      NaN   
1372                    N         1           NaN           NaN      NaN   
1776                    N         1           NaN           NaN      NaN   
1831                    N         1           NaN           NaN      NaN   
2244                    N         1           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1935651                 A         0           NaN           NaN      NaN   
1935716                 N         1           NaN           NaN      NaN   
1935876                 A         0           NaN           NaN      NaN   
1935978                 N         1           NaN           NaN      NaN   
1936470                 B         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1280              NaN                NaN  
1372              NaN                NaN  
1776              NaN                NaN  
1831              NaN                NaN  
2244              NaN                NaN  
...               ...                ...  
1935651           NaN                NaN  
1935716           NaN                NaN  
1935876           NaN                NaN  
1935978           NaN                NaN  
1936470           NaN                NaN  

[8387 rows x 30 columns]

In [28]:
cancelados_o_desviados = filas_nan[(filas_nan['Cancelled'] == 1) | (filas_nan['Diverted'] == 1)]
cancelados_o_desviados

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1280           1763  2008      1           3          4    922.0         915   
1372           1911  2008      1           3          4   2325.0        1900   
1776           2651  2008      1           4          5   1949.0        1905   
1831           2726  2008      1           4          5    737.0         705   
2244           3672  2008      1           4          5   1849.0        1630   
...             ...   ...    ...         ...        ...      ...         ...   
1935651     7006289  2008     12          10          3   1459.0        1447   
1935716     7006401  2008     12          11          4   1355.0        1106   
1935876     7006809  2008     12          11          4   1026.0         955   
1935978     7007034  2008     12          11          4   1527.0        1520   
1936470     7008584  2008     12          12          5    703.0         630   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1280         NaN        1050            WN  ...     NaN    12.0          0   
1372         NaN        2030            WN  ...     NaN    11.0          0   
1776         NaN        1910            WN  ...     NaN    11.0          0   
1831         NaN         825            WN  ...     NaN    13.0          0   
2244         NaN        1755            WN  ...     NaN    12.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1935651      NaN        1650            DL  ...     NaN    37.0          1   
1935716      7.0        1950            DL  ...    13.0    17.0          0   
1935876      NaN        1219            DL  ...     NaN     NaN          1   
1935978   2106.0        1708            DL  ...     9.0    19.0          0   
1936470      NaN         734            DL  ...     NaN    33.0          1   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1280                    N         1           NaN           NaN      NaN   
1372                    N         1           NaN           NaN      NaN   
1776                    N         1           NaN           NaN      NaN   
1831                    N         1           NaN           NaN      NaN   
2244                    N         1           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1935651                 A         0           NaN           NaN      NaN   
1935716                 N         1           NaN           NaN      NaN   
1935876                 A         0           NaN           NaN      NaN   
1935978                 N         1           NaN           NaN      NaN   
1936470                 B         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1280              NaN                NaN  
1372              NaN                NaN  
1776              NaN                NaN  
1831              NaN                NaN  
2244              NaN                NaN  
...               ...                ...  
1935651           NaN                NaN  
1935716           NaN                NaN  
1935876           NaN                NaN  
1935978           NaN                NaN  
1936470           NaN                NaN  

[8387 rows x 30 columns]

Coincide el número de valores NaN en AirTime y ArrDelay con los vuelos que han sido cancelados o desviados. En este casso puedo eliminar las filas que contengan NaN en estas variables.

In [30]:
airlines_delay = airlines_delay.drop(filas_nan.index)
airlines_delay

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1928371 rows x 30 columns]

Tambien elimino las filas con NaN en las variables que representan las causas de los retrasos. Estas filas corresponden a vuelos que no han teido retraso o ha sido de menos de 15 minutos. Elimino el resto de filas con valores NaN. Si no hay valor en ArrTime ni en TaxiIn no puedo verificar si hay retraso, ActualElapsedTime coincide con AirTime y ArrDelay, y el resto tienen pocos valores nulos con lo que creo que se pueden eliminar también al no perder mucha información. 

In [31]:
df_sin_nan = airlines_delay.dropna()

print(f"Filas eliminadas: {len(airlines_delay) - len(df_sin_nan)}")
df_sin_nan

Filas eliminadas: 680885


Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
3                 4  2008      1           3          4   1829.0        1755   
5                 6  2008      1           3          4   1937.0        1830   
7                11  2008      1           3          4   1644.0        1510   
9                16  2008      1           3          4   1452.0        1425   
11               18  2008      1           3          4   1323.0        1255   
...             ...   ...    ...         ...        ...      ...         ...   
1936751     7009705  2008     12          13          6    921.0         830   
1936752     7009709  2008     12          13          6   1552.0        1520   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
3         1959.0        1925            WN  ...     3.0    10.0          0   
5         2037.0        1940            WN  ...     3.0     7.0          0   
7         1845.0        1725            WN  ...     6.0     8.0          0   
9         1640.0        1625            WN  ...     7.0     8.0          0   
11        1526.0        1510            WN  ...     4.0     9.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936751   1112.0        1008            DL  ...     8.0    21.0          0   
1936752   1735.0        1718            DL  ...     9.0     7.0          0   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
3                       N         0           2.0           0.0      0.0   
5                       N         0          10.0           0.0      0.0   
7                       N         0           8.0           0.0      0.0   
9                       N         0           3.0           0.0      0.0   
11                      N         0           0.0           0.0      0.0   
...                   ...       ...           ...           ...      ...   
1936751                 N         0          51.0           0.0     13.0   
1936752                 N         0           0.0           0.0      0.0   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   

        SecurityDelay  LateAircraftDelay  
3                 0.0               32.0  
5                 0.0               47.0  
7                 0.0               72.0  
9                 0.0               12.0  
11                0.0               16.0  
...               ...                ...  
1936751           0.0                0.0  
1936752           0.0               17.0  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  

[1247486 rows x 30 columns]

### Eliminar columnas 

Creo que para este análisis podemos prescindir de la informacion sobre vuelos cancelados o desviados, el tiempo que pasa el avión en pista con los pasajeros dentro (TaxiIn, TaxiOut), la distancia y el destino del vuelo, el tiempo que ha pasado el avion en el aire, el número de vuelo y tambien la columna del año (todas son 2008) y unnamed.

In [32]:
columnas_para_eliminar = ['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 
                          'FlightNum', 'Cancelled', 'CancellationCode', 'Diverted', 'TaxiIn', 'TaxiOut']
airlines_delay_reducido = df_sin_nan.drop(columnas_para_eliminar, axis=1)
airlines_delay_reducido

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier TailNum  \
3         1829.0        1755   1959.0        1925            WN  N464WN   
5         1937.0        1830   2037.0        1940            WN  N763SW   
7         1644.0        1510   1845.0        1725            WN  N334SW   
9         1452.0        1425   1640.0        1625            WN  N286WN   
11        1323.0        1255   1526.0        1510            WN  N674AA   
...          ...         ...      ...         ...           ...     ...   
1936751    921.0         830   1112.0        1008            DL  N907DE   
1936752   1552.0        1520   1735.0        1718            DL  N905DE   
1936753   1250.0        1220   1617.0        1552            DL  N938DL   
1936754    657.0         600    904.0         749            DL  N3743H   
1936755   1007.0         847   1149.0        1010            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                     90.0            90.0     77.0      34.0      34.0   
5                    240.0           250.0    230.0      57.0      67.0   
7                    121.0           135.0    107.0      80.0      94.0   
9                    228.0           240.0    213.0      15.0      27.0   
11                   123.0           135.0    110.0      16.0      28.0   
...                    ...             ...      ...       ...       ...   
1936751              111.0            98.0     82.0      64.0      51.0   
1936752               43.0            58.0     27.0      17.0      32.0   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515           2.0           0.0       0.0   
5          IND  LAS      1591          10.0           0.0       0.0   
7          IND  MCO       828           8.0           0.0       0.0   
9          IND  PHX      1489           3.0           0.0       0.0   
11         IND  TPA       838           0.0           0.0       0.0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545          51.0           0.0      13.0   
1936752    HSV  ATL       151           0.0           0.0       0.0   
1936753    MSP  ATL       906           3.0           0.0       0.0   
1936754    RIC  ATL       481           0.0          57.0      18.0   
1936755    ATL  IAH       689           1.0           0.0      19.0   

         SecurityDelay  LateAircraftDelay  
3                  0.0               32.0  
5                  0.0               47.0  
7                  0.0               72.0  
9                  0.0               12.0  
11                 0.0               16.0  
...                ...                ...  
1936751            0.0                0.0  
1936752            0.0               17.0  
1936753            0.0               22.0  
1936754            0.0                0.0  
1936755            0.0               79.0  

[1247486 rows x 19 columns]

### Tipo de dato
Para unificar el dataframe y viendo que todos los decimales son .0, defino una función para covertrlo todo a int64.

In [33]:
def float_a_int(valor):
    if isinstance(valor, float):
        return int(valor)
    else:
        return valor

airlines_delay_reducido = airlines_delay_reducido.applymap(float_a_int)
airlines_delay_reducido

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier TailNum  \
3           1829        1755     1959        1925            WN  N464WN   
5           1937        1830     2037        1940            WN  N763SW   
7           1644        1510     1845        1725            WN  N334SW   
9           1452        1425     1640        1625            WN  N286WN   
11          1323        1255     1526        1510            WN  N674AA   
...          ...         ...      ...         ...           ...     ...   
1936751      921         830     1112        1008            DL  N907DE   
1936752     1552        1520     1735        1718            DL  N905DE   
1936753     1250        1220     1617        1552            DL  N938DL   
1936754      657         600      904         749            DL  N3743H   
1936755     1007         847     1149        1010            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                       90              90       77        34        34   
5                      240             250      230        57        67   
7                      121             135      107        80        94   
9                      228             240      213        15        27   
11                     123             135      110        16        28   
...                    ...             ...      ...       ...       ...   
1936751                111              98       82        64        51   
1936752                 43              58       27        17        32   
1936753                147             152      120        25        30   
1936754                127             109       78        75        57   
1936755                162             143      122        99        80   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515             2             0         0   
5          IND  LAS      1591            10             0         0   
7          IND  MCO       828             8             0         0   
9          IND  PHX      1489             3             0         0   
11         IND  TPA       838             0             0         0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545            51             0        13   
1936752    HSV  ATL       151             0             0         0   
1936753    MSP  ATL       906             3             0         0   
1936754    RIC  ATL       481             0            57        18   
1936755    ATL  IAH       689             1             0        19   

         SecurityDelay  LateAircraftDelay  
3                    0                 32  
5                    0                 47  
7                    0                 72  
9                    0                 12  
11                   0                 16  
...                ...                ...  
1936751              0                  0  
1936752              0                 17  
1936753              0                 22  
1936754              0                  0  
1936755              0                 79  

[1247486 rows x 19 columns]

In [34]:
columnas_horarias = ['DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime']

for columna in columnas_horarias:
    airlines_delay_reducido[columna] = pd.to_datetime(airlines_delay_reducido[columna], format='%H%M', errors='coerce')
for columna in columnas_horarias:
    airlines_delay_reducido[columna] = airlines_delay_reducido[columna].dt.strftime('%H:%M')

airlines_delay_reducido

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
3         18:29      17:55   19:59      19:25            WN  N464WN   
5         19:37      18:30   20:37      19:40            WN  N763SW   
7         16:44      15:10   18:45      17:25            WN  N334SW   
9         14:52      14:25   16:40      16:25            WN  N286WN   
11        13:23      12:55   15:26      15:10            WN  N674AA   
...         ...        ...     ...        ...           ...     ...   
1936751   09:21      08:30   11:12      10:08            DL  N907DE   
1936752   15:52      15:20   17:35      17:18            DL  N905DE   
1936753   12:50      12:20   16:17      15:52            DL  N938DL   
1936754   06:57      06:00   09:04      07:49            DL  N3743H   
1936755   10:07      08:47   11:49      10:10            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                       90              90       77        34        34   
5                      240             250      230        57        67   
7                      121             135      107        80        94   
9                      228             240      213        15        27   
11                     123             135      110        16        28   
...                    ...             ...      ...       ...       ...   
1936751                111              98       82        64        51   
1936752                 43              58       27        17        32   
1936753                147             152      120        25        30   
1936754                127             109       78        75        57   
1936755                162             143      122        99        80   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515             2             0         0   
5          IND  LAS      1591            10             0         0   
7          IND  MCO       828             8             0         0   
9          IND  PHX      1489             3             0         0   
11         IND  TPA       838             0             0         0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545            51             0        13   
1936752    HSV  ATL       151             0             0         0   
1936753    MSP  ATL       906             3             0         0   
1936754    RIC  ATL       481             0            57        18   
1936755    ATL  IAH       689             1             0        19   

         SecurityDelay  LateAircraftDelay  
3                    0                 32  
5                    0                 47  
7                    0                 72  
9                    0                 12  
11                   0                 16  
...                ...                ...  
1936751              0                  0  
1936752              0                 17  
1936753              0                 22  
1936754              0                  0  
1936755              0                 79  

[1247486 rows x 19 columns]

## Muestra aleatoria
Usamos una muestra aleatoria de 200.000 registros. Para asegurar que cada vez que se ejecute el código sea igual, utilizo el metodo random_state para fijar la semilla a partir de la cual genera 'muestra_airlines_delay'. Tenemos resultados numericos del tipo int64 y float64. Lo paso todo a float64 por la mayoria de columnas que ya tiene este tipo de dato.    

In [35]:
muestra_airlines_delay = airlines_delay_reducido.sample(n=200000, random_state=42)
muestra_airlines_delay

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
650090    17:11      16:35   18:29      17:45            WN  N717SA   
1472834   10:10      09:45   17:24      16:49            US  N667AW   
78226     17:57      15:32   20:26      18:09            UA  N349UA   
1826529   21:56      21:09   23:12      22:13            US  N822AW   
1463701   18:20      17:51   20:03      19:35            OO  N251YV   
...         ...        ...     ...        ...           ...     ...   
693087    15:54      15:30   18:31      18:00            AA  N519AA   
1116196   17:20      16:35   20:38      20:05            XE  N34110   
1158447   15:45      14:37   18:55      17:57            UA  N562UA   
1630534   18:04      17:10   19:27      18:35            WN  N292WN   
1398784   10:19      08:15   15:30      13:45            AA  N232AA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
650090                  78              70       57        44        36   
1472834                254             244      217        35        25   
78226                  149             157      117       137       145   
1826529                136             124      109        59        47   
1463701                103             104       93        28        29   
...                    ...             ...      ...       ...       ...   
693087                 217             210      183        31        24   
1116196                138             150      118        33        45   
1158447                130             140      109        58        68   
1630534                 83              85       70        52        54   
1398784                191             210      177       105       124   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
650090     BUR  OAK       325             0             0         8   
1472834    LAS  PIT      1910            25             0        10   
78226      SEA  LAX       954            73             0         0   
1826529    PHX  OAK       646             0             0        13   
1463701    OTH  SFO       412             0            28         0   
...        ...  ...       ...           ...           ...       ...   
693087     DCA  DFW      1192             0             9         7   
1116196    IAH  CVG       871            33             0         0   
1158447    LAX  DEN       862             5             0         0   
1630534    OAK  LAS       407             0             0         0   
1398784    SJC  AUS      1476            69             0         0   

         SecurityDelay  LateAircraftDelay  
650090               0                 36  
1472834              0                  0  
78226                0                 64  
1826529              0                 46  
1463701              0                  0  
...                ...                ...  
693087               0                 15  
1116196              0                  0  
1158447              0                 53  
1630534              0                 52  
1398784              0                 36  

[200000 rows x 19 columns]

## EXERCICI 2

### Fes un informe complet del dataset:

#### Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

In [36]:
# Excluyo las columnas cuyos resultados representan horas
columnas_a_excluir = ['DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime']
resumen_estadistico = muestra_airlines_delay.drop(columns=columnas_a_excluir).describe()

resumen_estadistico

ActualElapsedTime  CRSElapsedTime        AirTime       ArrDelay  \
count      200000.000000   200000.000000  200000.000000  200000.000000   
mean          135.592870      131.927200     107.577285      63.339455   
std            72.361889       69.623282      68.256659      60.965008   
min            17.000000        1.000000       0.000000      15.000000   
25%            83.000000       80.000000      58.000000      26.000000   
50%           118.000000      115.000000      90.000000      43.000000   
75%           167.000000      161.000000     136.000000      79.000000   
max           750.000000      660.000000     733.000000    1655.000000   

            DepDelay      Distance   CarrierDelay   WeatherDelay  \
count  200000.000000  200000.00000  200000.000000  200000.000000   
mean       59.673775     742.84279      19.273360       3.677815   
std        60.097999     560.07192      43.564064      21.482442   
min         6.000000      31.00000       0.000000       0.000000   
25%        24.000000     334.00000       0.000000       0.000000   
50%        41.000000     596.00000       2.000000       0.000000   
75%        75.000000     975.00000      21.000000       0.000000   
max      1597.000000    4962.00000    1351.000000    1297.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  200000.000000  200000.000000      200000.000000  
mean       14.992805       0.085845          25.309630  
std        33.629719       1.935511          42.318863  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         2.000000       0.000000           8.000000  
75%        15.000000       0.000000          33.000000  
max      1207.000000     193.000000        1236.000000

La función .describe() muestra un primer análisis estadístico. Podemos destacar que la media de la duración de los vuelos es de 135 minutos incluyemdo el tiempo en pista de los que de media está 107 minutos en el aire. Con esto vemos que el tiempo medio que pasa un avión en pista con los pasajeros sin moverse es de unos 28 minutos. La distancia media es de 742 millas. También vemos los percentiles y encuentro valores muy altos en max. He tomado como ejemplo ArrDelay pensando que quizá algún tiempo se habia quedado en el "formato hora" en lugar de en minutos, pero al realizar sort_values el siguiente resultado era 1592 con lo que no puede ser horas y minutos. También he examinado algunas de las filas con valores mas altos pero no he visto nada que me haya resultado extraño. No sé cuál es la razón de los valores tan altos en max.  

#### Troba quantes dades faltants hi ha per columna.


In [44]:
valores_nulos_por_columnas = muestra_airlines_delay.isnull().sum()
print (valores_nulos_por_columnas)

DepTime                 396
CRSDepTime                5
ArrTime                1781
CRSArrTime              522
UniqueCarrier             0
TailNum                   0
ActualElapsedTime         0
CRSElapsedTime            0
AirTime                   0
ArrDelay                  0
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
CarrierDelay              0
WeatherDelay              0
NASDelay                  0
SecurityDelay             0
LateAircraftDelay         0
Velocidad_Media_mph       0
Delayed                   0
dtype: int64


#### Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
Compruebo que la distancia está en millas y el tiempo en minutos. La columna muestra la velocidad media en Mph.


In [45]:
muestra_airlines_delay['Velocidad_Media_mph'] = muestra_airlines_delay['Distance'] / (muestra_airlines_delay['AirTime'] / 60)
muestra_airlines_delay

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
650090    17:11      16:35   18:29      17:45            WN  N717SA   
1472834   10:10      09:45   17:24      16:49            US  N667AW   
78226     17:57      15:32   20:26      18:09            UA  N349UA   
1826529   21:56      21:09   23:12      22:13            US  N822AW   
1463701   18:20      17:51   20:03      19:35            OO  N251YV   
...         ...        ...     ...        ...           ...     ...   
693087    15:54      15:30   18:31      18:00            AA  N519AA   
1116196   17:20      16:35   20:38      20:05            XE  N34110   
1158447   15:45      14:37   18:55      17:57            UA  N562UA   
1630534   18:04      17:10   19:27      18:35            WN  N292WN   
1398784   10:19      08:15   15:30      13:45            AA  N232AA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  Origin  \
650090                  78              70       57        44  ...     BUR   
1472834                254             244      217        35  ...     LAS   
78226                  149             157      117       137  ...     SEA   
1826529                136             124      109        59  ...     PHX   
1463701                103             104       93        28  ...     OTH   
...                    ...             ...      ...       ...  ...     ...   
693087                 217             210      183        31  ...     DCA   
1116196                138             150      118        33  ...     IAH   
1158447                130             140      109        58  ...     LAX   
1630534                 83              85       70        52  ...     OAK   
1398784                191             210      177       105  ...     SJC   

        Dest Distance  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
650090   OAK      325             0             0         8              0   
1472834  PIT     1910            25             0        10              0   
78226    LAX      954            73             0         0              0   
1826529  OAK      646             0             0        13              0   
1463701  SFO      412             0            28         0              0   
...      ...      ...           ...           ...       ...            ...   
693087   DFW     1192             0             9         7              0   
1116196  CVG      871            33             0         0              0   
1158447  DEN      862             5             0         0              0   
1630534  LAS      407             0             0         0              0   
1398784  AUS     1476            69             0         0              0   

         LateAircraftDelay  Velocidad_Media_mph  Delayed  
650090                  36           342.105263        1  
1472834                  0           528.110599        1  
78226                   64           489.230769        1  
1826529                 46           355.596330        1  
1463701                  0           265.806452        1  
...                    ...                  ...      ...  
693087                  15           390.819672        1  
1116196                  0           442.881356        1  
1158447                 53           474.495413        1  
1630534                 52           348.857143        1  
1398784                 36           500.338983        1  

[200000 rows x 21 columns]

In [46]:
muestra_airlines_delay['Delayed'] = muestra_airlines_delay['ArrTime'] > muestra_airlines_delay['CRSArrTime']
muestra_airlines_delay['Delayed'] = muestra_airlines_delay['Delayed'].astype(int)
muestra_airlines_delay

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
650090    17:11      16:35   18:29      17:45            WN  N717SA   
1472834   10:10      09:45   17:24      16:49            US  N667AW   
78226     17:57      15:32   20:26      18:09            UA  N349UA   
1826529   21:56      21:09   23:12      22:13            US  N822AW   
1463701   18:20      17:51   20:03      19:35            OO  N251YV   
...         ...        ...     ...        ...           ...     ...   
693087    15:54      15:30   18:31      18:00            AA  N519AA   
1116196   17:20      16:35   20:38      20:05            XE  N34110   
1158447   15:45      14:37   18:55      17:57            UA  N562UA   
1630534   18:04      17:10   19:27      18:35            WN  N292WN   
1398784   10:19      08:15   15:30      13:45            AA  N232AA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  Origin  \
650090                  78              70       57        44  ...     BUR   
1472834                254             244      217        35  ...     LAS   
78226                  149             157      117       137  ...     SEA   
1826529                136             124      109        59  ...     PHX   
1463701                103             104       93        28  ...     OTH   
...                    ...             ...      ...       ...  ...     ...   
693087                 217             210      183        31  ...     DCA   
1116196                138             150      118        33  ...     IAH   
1158447                130             140      109        58  ...     LAX   
1630534                 83              85       70        52  ...     OAK   
1398784                191             210      177       105  ...     SJC   

        Dest Distance  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
650090   OAK      325             0             0         8              0   
1472834  PIT     1910            25             0        10              0   
78226    LAX      954            73             0         0              0   
1826529  OAK      646             0             0        13              0   
1463701  SFO      412             0            28         0              0   
...      ...      ...           ...           ...       ...            ...   
693087   DFW     1192             0             9         7              0   
1116196  CVG      871            33             0         0              0   
1158447  DEN      862             5             0         0              0   
1630534  LAS      407             0             0         0              0   
1398784  AUS     1476            69             0         0              0   

         LateAircraftDelay  Velocidad_Media_mph  Delayed  
650090                  36           342.105263        1  
1472834                  0           528.110599        1  
78226                   64           489.230769        1  
1826529                 46           355.596330        1  
1463701                  0           265.806452        1  
...                    ...                  ...      ...  
693087                  15           390.819672        1  
1116196                  0           442.881356        1  
1158447                 53           474.495413        1  
1630534                 52           348.857143        1  
1398784                 36           500.338983        1  

[200000 rows x 21 columns]

In [47]:
vuelos_retrasados = muestra_airlines_delay['Delayed'].value_counts()

print(vuelos_retrasados)

1    187464
0     12536
Name: Delayed, dtype: int64


La columna Delayed muestra si el vuelo se ha retrasado (1) o no (0). En la muestra aleatoria hay un total de 188205 vuelos retrasados.

#### Fes una taula de les aerolínies amb més endarreriments acumulats.

Conteo de vuelos por cada aerolinea. A continuación se muestra el número de retrasos de cada una de ellas y el porcentaje.

In [48]:
num_aerolineas = muestra_airlines_delay['UniqueCarrier'].value_counts()
num_retrasos = muestra_airlines_delay[muestra_airlines_delay['Delayed'] > 0].groupby('UniqueCarrier')['Delayed'].count()
resultados_combinados = pd.DataFrame({'VuelosTotales': num_aerolineas, 'Retrasos': num_retrasos})
resultados_combinados['PorcentajeRetrasos'] = (resultados_combinados['Retrasos'] / resultados_combinados['VuelosTotales']) * 100
resultados_combinados = resultados_combinados.sort_values(by='VuelosTotales', ascending=False)

print(resultados_combinados)

    VuelosTotales  Retrasos  PorcentajeRetrasos
WN          32523     30703           94.403960
AA          21242     19577           92.161755
MQ          15596     15129           97.005642
UA          15165     13945           91.955160
OO          14289     13655           95.563021
DL          11690     10905           93.284859
XE          11588     10926           94.287194
US           9556      8893           93.061951
CO           9483      8558           90.245703
EV           9172      8869           96.696468
NW           8674      8193           94.454692
YV           8076      7577           93.821199
FL           7567      7052           93.194132
OH           6319      5995           94.872606
B6           6039      5164           85.510846
9E           5728      5477           95.618017
AS           3935      3746           95.196950
F9           2624      2408           91.768293
HA            672       631           93.898810
AQ             62        61           98

#### Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

In [49]:
rutas = muestra_airlines_delay.sort_values(by='Distance', ascending=False)
rutas_mas_largas = rutas[['Origin', 'Dest', 'Distance']].drop_duplicates(subset='Distance')

print(rutas_mas_largas)

        Origin Dest  Distance
1623374    EWR  HNL      4962
1078905    ATL  HNL      4502
1323556    HNL  ORD      4243
956585     KOA  ORD      4213
1147091    ORD  OGG      4184
...        ...  ...       ...
620015     OXR  LAX        49
53161      LAX  ONT        47
1240266    GST  JNU        41
54674      LAX  SNA        36
1060338    PSG  WRG        31

[1370 rows x 3 columns]


La ruta más larga es de Newark (EWR) a Honolulu (HNL).

In [50]:
retrasos_acumulados = muestra_airlines_delay.groupby(['Origin', 'Dest'])['Delayed'].sum().reset_index()
rutas_con_mas_retrasos = retrasos_acumulados.sort_values(by='Delayed', ascending=False)

print(rutas_con_mas_retrasos)

     Origin Dest  Delayed
2523    LAX  SFO      487
3479    ORD  LGA      464
210     ATL  LGA      441
4272    SFO  LAX      430
2597    LGA  ORD      407
...     ...  ...      ...
2354    KOA  SAN        0
2419    LAS  MLI        0
2052    IAH  BDL        0
4406    SLC  FAI        0
2582    LGA  HSV        0

[4812 rows x 3 columns]


La ruta que más retrasos acumula es Los Angeles (LAX) San Francisco (SFO).

In [53]:
muestra_airlines_delay.to_excel('estructures_Dataframe.xlsx', index=False)